axoproject/src/lib.rs

```diff
@@ -33,6 +33,186 @@ mod tests;
+    }
+
+    /// Get the root workspace's index
+    pub fn root_workspace_idx(&self) -> WorkspaceIdx {
+        WorkspaceIdx(0)
+    }
+    /// Get the root workspace
+    pub fn root_workspace(&self) -> &WorkspaceInfo {
+        self.workspace(self.root_workspace_idx())
+    }
+
+    /// Get a workspace
+    pub fn workspace(&self, idx: WorkspaceIdx) -> &WorkspaceInfo {
+        &self.workspaces[idx.0]
+    }
+    /// Get a package
+    pub fn package(&self, idx: PackageIdx) -> &PackageInfo {
+        &self.packages[idx.0]
+    }
+    /// Get a mutable workspace
+    pub fn workspace_mut(&mut self, idx: WorkspaceIdx) -> &mut WorkspaceInfo {
+        &mut self.workspaces[idx.0]
+    }
+    /// Get a mutable package
+    pub fn package_mut(&mut self, idx: PackageIdx) -> &mut PackageInfo {
+        &mut self.packages[idx.0]
+    }
+    /// Get the parent workspace of a package
+    pub fn workspace_for_package(&self, idx: PackageIdx) -> WorkspaceIdx {
+        self.package_parents[&idx]
+    }
+
+    /// Get a workspace's packages (only direct children)
+    pub fn direct_packages(
+        &self,
+        idx: WorkspaceIdx,
+    ) -> impl Iterator<Item = (PackageIdx, &PackageInfo)> {
+        self.workspace_package_children
+            .get(&idx)
+            .map(|pkgs| &**pkgs)
+            .unwrap_or_default()
+            .iter()
+            .map(|p_idx| (*p_idx, &self.packages[p_idx.0]))
+    }
+
+    /// Get a workspace's packages (recursively getting children of subworkspaces)
+    pub fn recursive_packages(
+        &self,
+        idx: WorkspaceIdx,
+    ) -> impl Iterator<Item = (PackageIdx, &PackageInfo)> {
+        let mut working_set = vec![idx];
+        let mut package_indices = vec![];
+        while let Some(workspace_idx) = working_set.pop() {
+            // Add own packages
+            if let Some(packages) = self.workspace_package_children.get(&workspace_idx) {
+                package_indices.extend(packages.iter().copied());
+            }
+            // Add child workspaces to working set
+            if let Some(sub_workspaces) = self.workspace_workspace_children.get(&workspace_idx) {
+                working_set.extend(sub_workspaces.iter().copied());
+            }
+        }
+        // Return result
+        package_indices
+            .into_iter()
+            .map(|idx| (idx, self.package(idx)))
+    }
+
+    /// Get all packages
+    pub fn all_packages(&self) -> impl Iterator<Item = (PackageIdx, &PackageInfo)> {
+        self.packages
+            .iter()
+            .enumerate()
+            .map(|(idx, pkg)| (PackageIdx(idx), pkg))
+    }
+
+    /// Get all workspaces
+    pub fn all_workspace_indices(&self) -> impl Iterator<Item = WorkspaceIdx> {
+        (0..self.workspaces.len()).map(WorkspaceIdx)
+    }
+
+    /// Try to get a consensus repository_url for the given packages (or all if None given)
+    pub fn repository_url(&self, packages: Option<&[PackageIdx]>) -> Result<Option<RepositoryUrl>> {
```

Comment:
Wondering if it also makes sense to keep a version of this on individual workspaces. Guess we're probably not looking at that individually as often.

Reply:
Yeah you can always get that by just querying all packages under the workspace.